src/node.rs

```diff
@@ -0,0 +1,153 @@
+use crate::helper::Identifier;
+use crate::kbucket::{Bucket, KbucketTable, TableRecord};
+use std::collections::HashMap;
+use std::net::{Ipv4Addr, SocketAddrV4};
+use tokio::io;
+use tokio::net::UdpSocket;
+
+// The main Kademlia client struct.
+// Provides user-level API for performing querie and interacting with the underlying service.
+#[derive(Debug, PartialEq)]
+pub struct Node {
+    pub node_id: Identifier,
+    pub table: KbucketTable,
+    pub store: HashMap<Vec<u8>, Vec<u8>>, // Same storage as Portal network to store samples
+}
+
+impl Node {
+    pub fn new(node_id: Identifier, table_record: TableRecord) -> Self {
+        Self {
+            node_id,
+            table: KbucketTable::new(node_id, table_record),
+            store: Default::default(),
+        }
+    }
+
+    // Protocol's RPCs:
+    // ---------------------------------------------------------------------------------------------------
+    /// "The most important procedure a Kademlia participant must perform is to locate
+    /// the k closest nodes to some given node ID"
+    ///     - Kademlia Paper
+    ///
+    /// Recieves an id request and returns node information on nodes within
+    /// *its closest bucket* (instead of k-closest nodes) to that id.
+    pub fn find_node(&mut self, id: &Identifier) -> Vec<TableRecord> {
+        let bucket_index = self.table.xor_bucket_index(id);
+        self.table.buckets[bucket_index]
+            .map
+            .values()
+            .cloned()
+            .collect()
+    }
+
+    pub async fn ping(&mut self, local_socket: &UdpSocket, node_to_ping: &Identifier) -> usize {
+        let message_packet = b"Ping";
+
+        match self.table.get(node_to_ping) {
+            Some(remote_record) => {
+                let remote_socket =
+                    SocketAddrV4::new(remote_record.ip_address, remote_record.udp_port);
+                local_socket.connect(remote_socket).await;
+                local_socket.send(message_packet).await.unwrap()
+            }
+            _ => unreachable!("Node wasn't found to ping"),
+        }
+    }
+
+    // TODO:
+    // pub fn find_value() {}
+
+    /// "Most operations are implemented in terms of the lookup proceedure. To store a
+    /// <key,value> pair, a participant locates the k closes nodes to the key and sends them store RPCs".
+    ///
+    // TODO: Instructs a node to store a key, value pair for later retrieval.
+    // pub fn store(&mut self, key: Identifier, value: Vec<u8>) {}
+
+    // ---------------------------------------------------------------------------------------------------
+    pub async fn socket(&self) -> io::Result<UdpSocket> {
+        let table_record = self.table.local_record;
+        let socket_addr = SocketAddrV4::new(table_record.ip_address, table_record.udp_port);
+        let socket = UdpSocket::bind(socket_addr).await;
+        socket
```

Comment:
you could at least make the local `socket_addr` in `Node::new` and then not need to both with the `table` at all here

I try to avoid doing heavy work in constructors like `new` but you could also consider just binding the local socket and having it sit for the node's lifetime